infrastructure-playbooks/purge-docker-cluster.yml

```diff
@@ -507,25 +507,18 @@
 
       - name: remove pip and docker on debian
         apt:
-          name: "{{ item }}"
+          name: "{{ python-pip | union(docker-engine) }}"
```

Comment:
This won't work as expected I think.
Two things:
1/ you would have to add quotes --> `'python-pip' | union('docker-engine')`
2/ In anycase, I'm not sure this is the right approach. You should simply pass it this way: `['python-pip', 'docker-engine']`

Reply:
(You can also create a variable for this)

Reply:
Guillaume Abrioux [wrote](https://github.com/ceph/ceph-ansible/pull/3295#discussion_r230668366):
> This won't work as expected I think.

It did work for me; I tested it before creating the PR. 

The playbook - 
```
---
- hosts: localhost
  gather_facts: false
  tasks:
    - set_fact:
        list_one:
          - nano

    - set_fact:
        list_two:
          - python-kiwi-glade.x86_64
          - python-kiwi-docs.x86_64
    - name:
      dnf:
        name: "{{ list_one | union(list_two) }}"
        state: present
      become: true
```

The output -
```
$ sudo ansible-playbook -vvv jinja2-filters.yml | tee output.log
ansible-playbook 2.6.5
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible-playbook
  python version = 2.7.15 (default, May  9 2018, 11:18:37) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
Using /etc/ansible/ansible.cfg as config file
Parsed /etc/ansible/hosts inventory source with ini plugin
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAYBOOK: jinja2-filters.yml ********************************************************************************************************************************************************************
1 plays in jinja2-filters.yml

PLAY [localhost] ********************************************************************************************************************************************************************************
META: ran handlers

TASK [set_fact] *********************************************************************************************************************************************************************************
task path: /home/rishabh/temp/ansible/jinja2-filters.yml:5
ok: [localhost] => {
    "ansible_facts": {
        "list_one": [
            "nano"
        ]
    }, 
    "changed": false
}

TASK [set_fact] *********************************************************************************************************************************************************************************
task path: /home/rishabh/temp/ansible/jinja2-filters.yml:9
ok: [localhost] => {
    "ansible_facts": {
        "list_two": [
            "python-kiwi-glade.x86_64", 
            "python-kiwi-docs.x86_64"
        ]
    }, 
    "changed": false
}

TASK [dnf] **************************************************************************************************************************************************************************************
task path: /home/rishabh/temp/ansible/jinja2-filters.yml:13
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: rishabh
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1541411338.07-182687161836800 `" && echo ansible-tmp-1541411338.07-182687161836800="` echo /root/.ansible/tmp/ansible-tmp-1541411338.07-182687161836800 `" ) && sleep 0'
Using module file /usr/lib/python2.7/site-packages/ansible/modules/packaging/os/dnf.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-9644uHTqjg/tmpjbaf1_ TO /root/.ansible/tmp/ansible-tmp-1541411338.07-182687161836800/dnf.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1541411338.07-182687161836800/ /root/.ansible/tmp/ansible-tmp-1541411338.07-182687161836800/dnf.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-thicpzkfjdledzhspjldcatawuglehff; /usr/bin/python2 /root/.ansible/tmp/ansible-tmp-1541411338.07-182687161836800/dnf.py'"'"' && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1541411338.07-182687161836800/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true, 
    "invocation": {
        "module_args": {
            "autoremove": null, 
            "conf_file": null, 
            "disable_gpg_check": false, 
            "disablerepo": [], 
            "enablerepo": [], 
            "installroot": "/", 
            "list": null, 
            "name": [
                "nano", 
                "python-kiwi-glade.x86_64", 
                "python-kiwi-docs.x86_64"
            ], 
            "releasever": null, 
            "state": "present"
        }
    }, 
    "results": [
        "Installed: python-kiwi-docs-1.9.38-5.fc23.x86_64", 
        "Installed: nano-2.8.7-1.fc27.x86_64", 
        "Installed: python-kiwi-glade-1.9.38-5.fc23.x86_64", 
        "Installed: glade-3.20.4-1.fc27.x86_64", 
        "Installed: glade-libs-3.20.4-1.fc27.x86_64", 
        "Installed: python-kiwi-1.9.38-5.fc23.x86_64"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP **************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0
```

Reply:
And I tried your way too, but it didn't work for me -

The playbook -
```
---
- hosts: localhost
  gather_facts: false
  tasks:
    - set_fact:
        list_one:
          - nano

    - set_fact:
        list_two:
          - python-kiwi-glade.x86_64
          - python-kiwi-docs.x86_64
    - name:
      dnf:
        name: "{{ ['list_one', 'list_two'] }}"
        state: present
      become: true
```

Output
```
$ ansible-playbook -vvv jinja2-filters.yml 
ansible-playbook 2.7.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/rishabh/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/rishabh/.local/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.15 (default, May  9 2018, 11:18:37) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
Using /etc/ansible/ansible.cfg as config file
Parsed /etc/ansible/hosts inventory source with ini plugin
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAYBOOK: jinja2-filters.yml ********************************************************************************************************************************************************************
1 plays in jinja2-filters.yml

PLAY [localhost] ********************************************************************************************************************************************************************************
META: ran handlers

TASK [set_fact] *********************************************************************************************************************************************************************************
task path: /home/rishabh/temp/ansible/jinja2-filters.yml:5
ok: [localhost] => {
    "ansible_facts": {
        "list_one": [
            "nano"
        ]
    }, 
    "changed": false
}

TASK [set_fact] *********************************************************************************************************************************************************************************
task path: /home/rishabh/temp/ansible/jinja2-filters.yml:9
ok: [localhost] => {
    "ansible_facts": {
        "list_two": [
            "python-kiwi-glade.x86_64", 
            "python-kiwi-docs.x86_64"
        ]
    }, 
    "changed": false
}

TASK [dnf] **************************************************************************************************************************************************************************************
task path: /home/rishabh/temp/ansible/jinja2-filters.yml:13
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: rishabh
<127.0.0.1> EXEC /bin/sh -c 'echo ~rishabh && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/rishabh/.ansible/tmp/ansible-tmp-1541411800.53-133115250721315 `" && echo ansible-tmp-1541411800.53-133115250721315="` echo /home/rishabh/.ansible/tmp/ansible-tmp-1541411800.53-133115250721315 `" ) && sleep 0'
Using module file /home/rishabh/.local/lib/python2.7/site-packages/ansible/modules/packaging/os/dnf.py
<127.0.0.1> PUT /home/rishabh/.ansible/tmp/ansible-local-10474Mcndyg/tmpN7bsGe TO /home/rishabh/.ansible/tmp/ansible-tmp-1541411800.53-133115250721315/AnsiballZ_dnf.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/rishabh/.ansible/tmp/ansible-tmp-1541411800.53-133115250721315/ /home/rishabh/.ansible/tmp/ansible-tmp-1541411800.53-133115250721315/AnsiballZ_dnf.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-ptlvdojjbihwppeaxadfqzwfpudchqbb; /usr/bin/python2 /home/rishabh/.ansible/tmp/ansible-tmp-1541411800.53-133115250721315/AnsiballZ_dnf.py'"'"' && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/rishabh/.ansible/tmp/ansible-tmp-1541411800.53-133115250721315/ > /dev/null 2>&1 && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "module_stderr": "sudo: a password is required\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", 
    "rc": 1
}
	to retry, use: --limit @/home/rishabh/temp/ansible/jinja2-filters.retry

PLAY RECAP **************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1   

[rishabh@x220 ansible]$ sudo ansible-playbook -vvv jinja2-filters.yml 
[sudo] password for rishabh: 
ansible-playbook 2.6.5
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible-playbook
  python version = 2.7.15 (default, May  9 2018, 11:18:37) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
Using /etc/ansible/ansible.cfg as config file
Parsed /etc/ansible/hosts inventory source with ini plugin
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAYBOOK: jinja2-filters.yml ********************************************************************************************************************************************************************
1 plays in jinja2-filters.yml

PLAY [localhost] ********************************************************************************************************************************************************************************
META: ran handlers

TASK [set_fact] *********************************************************************************************************************************************************************************
task path: /home/rishabh/temp/ansible/jinja2-filters.yml:5
ok: [localhost] => {
    "ansible_facts": {
        "list_one": [
            "nano"
        ]
    }, 
    "changed": false
}

TASK [set_fact] *********************************************************************************************************************************************************************************
task path: /home/rishabh/temp/ansible/jinja2-filters.yml:9
ok: [localhost] => {
    "ansible_facts": {
        "list_two": [
            "python-kiwi-glade.x86_64", 
            "python-kiwi-docs.x86_64"
        ]
    }, 
    "changed": false
}

TASK [dnf] **************************************************************************************************************************************************************************************
task path: /home/rishabh/temp/ansible/jinja2-filters.yml:13
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: rishabh
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1541411811.91-67083974285820 `" && echo ansible-tmp-1541411811.91-67083974285820="` echo /root/.ansible/tmp/ansible-tmp-1541411811.91-67083974285820 `" ) && sleep 0'
Using module file /usr/lib/python2.7/site-packages/ansible/modules/packaging/os/dnf.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-10512TMPmvi/tmp9u2zFG TO /root/.ansible/tmp/ansible-tmp-1541411811.91-67083974285820/dnf.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1541411811.91-67083974285820/ /root/.ansible/tmp/ansible-tmp-1541411811.91-67083974285820/dnf.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-zyndulqmqnqgyjundosrxtodftoefylv; /usr/bin/python2 /root/.ansible/tmp/ansible-tmp-1541411811.91-67083974285820/dnf.py'"'"' && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1541411811.91-67083974285820/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
  File "/tmp/ansible_u4yCAL/ansible_module_dnf.py", line 293, in _mark_package_install
    base.install(pkg_spec)
  File "/usr/lib/python2.7/site-packages/dnf/base.py", line 1720, in install
    raise dnf.exceptions.MarkingError(_('no package matched'), pkg_spec)

fatal: [localhost]: FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "autoremove": null, 
            "conf_file": null, 
            "disable_gpg_check": false, 
            "disablerepo": [], 
            "enablerepo": [], 
            "installroot": "/", 
            "list": null, 
            "name": [
                "list_one", 
                "list_two"
            ], 
            "releasever": null, 
            "state": "present"
        }
    }, 
    "msg": "No package list_one available."
}
	to retry, use: --limit @/home/rishabh/temp/ansible/jinja2-filters.retry

PLAY RECAP **************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1  
```

Reply:
Yes, I expect the following to work :
```
---
- hosts: localhost
  gather_facts: false
  tasks:
    - set_fact:
        list_one:
          - nano

    - set_fact:
        list_two:
          - python-kiwi-glade.x86_64
          - python-kiwi-docs.x86_64
    - name:
      dnf:
        name: "{{ ['list_one', 'list_two'] }}"
        state: present
      become: true
```

indeed, `list_one` and `list_two` are both array type.

I don't expect this to work :
```
"{{ python-pip | union(docker-engine) }}"
```

because, `python-pip` as written in your patch are referring to a variable name that won't be defined since there is no quotes. If you add quotes anyway, it will result by returning something like : `[ 'p', 'y', 't', 'h', 'o', 'n', '-', 'p', 'i', 'p', 'd', 'o', 'c', 'k', 'e', 'r', '-', 'e', 'n', 'g', 'i', 'n', 'e' ]`


The way you tested my approach didn't work because you should have done this instead :

```
    - name:
      dnf:
        name: "{{ [list_one, list_two] }}"
        state: present
      become: true
```

Reply:
But I _think_ I see my mistake now: I've mistaken `docker-engine`, `docker` and `python-pip` as a variables holding  a list of packages but in reality they are package names.  :P

I will make changes that look like the following -

The playbook -
```
---
- hosts: localhost
  gather_facts: false
  tasks:
    - name:
      dnf:
        name: [nano, python-kiwi-glade.x86_64, python-kiwi-docs.x86_64]
        state: present
      become: true
```

The output -
```
$ sudo ansible-playbook -vvv jinja2-filters.yml 
ansible-playbook 2.6.5
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible-playbook
  python version = 2.7.15 (default, May  9 2018, 11:18:37) [GCC 7.3.1 20180303 (Red Hat 7.3.1-5)]
Using /etc/ansible/ansible.cfg as config file
Parsed /etc/ansible/hosts inventory source with ini plugin
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAYBOOK: jinja2-filters.yml ********************************************************************************************************************************************************************
1 plays in jinja2-filters.yml

PLAY [localhost] ********************************************************************************************************************************************************************************
META: ran handlers

TASK [dnf] **************************************************************************************************************************************************************************************
task path: /home/rishabh/temp/ansible/jinja2-filters.yml:5
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: rishabh
<127.0.0.1> EXEC /bin/sh -c 'echo ~root && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp/ansible-tmp-1541417315.98-176370320201664 `" && echo ansible-tmp-1541417315.98-176370320201664="` echo /root/.ansible/tmp/ansible-tmp-1541417315.98-176370320201664 `" ) && sleep 0'
Using module file /usr/lib/python2.7/site-packages/ansible/modules/packaging/os/dnf.py
<127.0.0.1> PUT /root/.ansible/tmp/ansible-local-13279L0lBYM/tmpkRm1CQ TO /root/.ansible/tmp/ansible-tmp-1541417315.98-176370320201664/dnf.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1541417315.98-176370320201664/ /root/.ansible/tmp/ansible-tmp-1541417315.98-176370320201664/dnf.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-tebialnhnfjknzdzwicjdxpaqjrvamff; /usr/bin/python2 /root/.ansible/tmp/ansible-tmp-1541417315.98-176370320201664/dnf.py'"'"' && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1541417315.98-176370320201664/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => {
    "changed": true, 
    "invocation": {
        "module_args": {
            "autoremove": null, 
            "conf_file": null, 
            "disable_gpg_check": false, 
            "disablerepo": [], 
            "enablerepo": [], 
            "installroot": "/", 
            "list": null, 
            "name": [
                "nano", 
                "python-kiwi-glade.x86_64", 
                "python-kiwi-docs.x86_64"
            ], 
            "releasever": null, 
            "state": "present"
        }
    }, 
    "results": [
        "Installed: python-kiwi-docs-1.9.38-5.fc23.x86_64", 
        "Installed: nano-2.8.7-1.fc27.x86_64", 
        "Installed: python-kiwi-glade-1.9.38-5.fc23.x86_64", 
        "Installed: glade-3.20.4-1.fc27.x86_64", 
        "Installed: glade-libs-3.20.4-1.fc27.x86_64", 
        "Installed: python-kiwi-1.9.38-5.fc23.x86_64"
    ]
}
META: ran handlers
META: ran handlers

PLAY RECAP **************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0 
```

Reply:
@rishabh-d-dave yes, sorry, I think it's :

```
name:
dnf:
name: "{{ list_one + list_two }}"
state: present
become: true
```